AFQ/data.py

```diff
@@ -288,6 +295,1039 @@ def read_templates(resample_to=False):
+
+            def get_subs_from_tsv_key(s3_key):
+                response = s3.get_object(
+                    Bucket=self.bucket,
+                    Key=s3_key
+                )
+
+                return set(pd.read_csv(
+                    response.get('Body')
+                ).participant_id.values)
+
+            subject_set = get_subs_from_tsv_key(s3_key)
+            subjects = list(subject_set)
+        else:
+            s3_prefix = '/'.join([self.bucket, self.s3_prefix])
+            sub_keys = _ls_s3fs(s3_prefix=s3_prefix,
+                                anon=self.anon)['subjects']
+
+            # Just need BIDSLayout for the `parse_file_entities`
+            # method so we can pass dev/null as the argument
+            layout = BIDSLayout(os.devnull, validate=False)
+            subjects = []
+            for key in sub_keys:
+                entities = layout.parse_file_entities(key)
+                subjects.append('sub-' + entities.get('subject'))
+
+        return subjects
+
+    def _download_non_sub_keys(self, directory, select="all"):
+        fs = s3fs.S3FileSystem(anon=self.anon)
+        for fn in self.non_sub_s3_keys['raw']:
+            if select == "all" or any([s in fn for s in select]):
+                fs.get(fn, op.join(directory, op.basename(fn)))
+
+    def download(self, directory, include_modality_agnostic=False,
+                 include_dataset_description=True,
+                 include_derivs=False, overwrite=False, pbar=True):
+        """Download files for each subject in the study
+
+        Parameters
+        ----------
+        directory : str
+            Directory to which to download subject files
+
+        include_modality_agnostic : bool, "all" or any subset of ["dataset_description.json", "CHANGES", "README", "LICENSE"]
+            If True or "all", download all keys in self.non_sub_s3_keys
+            also. If a subset of ["dataset_description.json", "CHANGES",
+            "README", "LICENSE"], download only those files. This is
+            useful if the non_sub_s3_keys contain files common to all
+            subjects that should be inherited. Default: False
+
+        include_derivs : bool or str
+            If True, download all derivatives files. If False, do not.
+            If a string or sequence of strings is passed, this will
+            only download derivatives that match the string(s) (e.g.
+            ["dmriprep", "afq"]). Default: False
+
+        overwrite : bool
+            If True, overwrite files for each subject. Default: False
+
+        pbar : bool
+            If True, include progress bar. Default: True
+
+        See Also
+        --------
+        AFQ.data.S3BIDSSubject.download
+        """
+        self._local_directories.append(directory)
+        self._local_directories = set(self._local_directories)
+
+        if include_modality_agnostic is True or include_modality_agnostic == "all":
+            _download_non_sub_keys(directory, select="all")
```

Comment:
This should be self._download_non_sub_keys I believe, as well as the one below.

Reply:
Yep, thanks.